## Python modules

* A Python module is a file containing Python code, including functions, variables, and classes, that can be reused in other Python programs.
* Modules help organize code into manageable pieces.
* Python provides many built-in modules (e.g., math, os, random), and
* you can also create user-defined modules by saving Python code in a **.py** file.
* To use a module, you import it into your script using the import statement.
* In general, any file with .py extension is treated as a module
* If the file is run as standalone script, it is treated as ‘main’ module.
* When the variables, classes and functions within a file are required in another python script, the file can be imported into that script using the syntax:
  - import\<space>filename
* Normally,  a collection of general purpose functions and classes are written a single python file and will be given a module name.
* This module can be imported and can be used in any python script
* Modules can be either built in or user defined
* Modules  as the name suggests,  play key role in modular programming.


### How python interpreter searches for modules:
* First it looks for the module (.py file) in the current working directory
* Then it will look into the system PATH (sys.path)

In [9]:
import sys
sys.path          # path is a variable of sys module. Accessed similarly like a method using a ' . ' (dot) operator
#sys.getprofile()  # a function in sys module, accessed in a same way

['C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_3.12.1776.0_x64__qbz5n2kfra8p0\\python312.zip',
 'C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_3.12.1776.0_x64__qbz5n2kfra8p0\\DLLs',
 'C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_3.12.1776.0_x64__qbz5n2kfra8p0\\Lib',
 'C:\\Program Files\\WindowsApps\\PythonSoftwareFoundation.Python.3.12_3.12.1776.0_x64__qbz5n2kfra8p0',
 '',
 'C:\\Users\\shiva\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages',
 'C:\\Users\\shiva\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\win32',
 'C:\\Users\\shiva\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python312\\site-packages\\win32\\lib',
 'C:\\Users\\shiva\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Pyth

* if module not found in those places, interpreter will throw an error

In [10]:
dir(sys)

['__breakpointhook__',
 '__displayhook__',
 '__doc__',
 '__excepthook__',
 '__interactivehook__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '__stderr__',
 '__stdin__',
 '__stdout__',
 '__unraisablehook__',
 '_base_executable',
 '_clear_type_cache',
 '_current_exceptions',
 '_current_frames',
 '_debugmallocstats',
 '_enablelegacywindowsfsencoding',
 '_framework',
 '_getframe',
 '_getframemodulename',
 '_git',
 '_home',
 '_setprofileallthreads',
 '_settraceallthreads',
 '_stdlib_dir',
 '_vpath',
 '_xoptions',
 'activate_stack_trampoline',
 'addaudithook',
 'api_version',
 'argv',
 'audit',
 'base_exec_prefix',
 'base_prefix',
 'breakpointhook',
 'builtin_module_names',
 'byteorder',
 'call_tracing',
 'copyright',
 'deactivate_stack_trampoline',
 'displayhook',
 'dllhandle',
 'dont_write_bytecode',
 'exc_info',
 'excepthook',
 'exception',
 'exec_prefix',
 'executable',
 'exit',
 'flags',
 'float_info',
 'float_repr_style',
 'get_asyncgen_hooks',
 'get_coroutine_origin_tra

### Using if `__name__ == "__main__"`

* Any **.py** file in python could be run directly or can be imported by another script.
* the expression if `__name__ == "__main__"`: is used to determine whether a script is being run directly (standalone) or imported as a module into another script.
* **`__name__:`** Every Python module (file) has a built-in variable called `__name__`. When a script is run directly, the interpreter sets `__name__` to `"__main__"`.
* However, if the script is imported as a module in another script, `__name__` will be set to the module's name (the file name without the .py extension).
* **`"__main__":`** This is a string that the Python interpreter assigns to `__name__` when the script is executed directly 

In [23]:
# my_module.py
def greet():
    print("Hello from the greet function!")

if __name__ == "__main__":
    # This block will only execute when this script is run directly
    print("Script is being run directly")
    greet()
else:
    # This block will execute when the module is imported
    print("Script has been imported")


Script is being run directly
Hello from the greet function!


In [50]:
%run my_module.py

Script is being run directly
Hello from the greet function!


### Importing module

* `import myModule`
* `import myModule as mym`
* The keyword ‘as’ helps to change the name of the module to ‘mym’ and thereby avoids any conflict with identifiers used in that particular script
* The statement import `module_name` only places `module_name` in the caller’s `symbol table`. 
* The objects that are defined in the module remain in the module’s private symbol table.
* To access the variable and methods of mym use **.** operator.
* E.g.:
  - `mym.some_method()`
  - `mym.some_variable`

* To import all functions and objects to the present script, we need to use the syntax
  - `from modulename import *`
* To import a specific function, we can use the syntax
  - `from modulename import function name`
  - `from modulename import function name as new_name`
* Examples:
  - `from math import *`
  - `from math import pow as p, sqrt as sq`


### Symbol table

* Each python script carries a finite number of identifiers
* When the script is being interpreted, python generates a symbol table containing the details of all the identifiers related to that script
* The symbol table will comprise of the name given by the user, the type of the identifier, its scope and its location in the memory
* The names of identifiers available in the local symbol table can be find out using dir()


### Scope and Namespace in Python

* In Python, **scope and namespace** are concepts that define the visibility and accessibility of variables within different parts of a program.

#### 1. Namespace:
* A namespace is a container that holds names (variables, functions, classes, etc.) as **keys** and their associated objects as **values**.
* In Python, different namespaces exist to avoid naming conflicts between variables or functions.
* For instance, two different functions can have variables with the same name, but since they exist in different namespaces, they won't conflict.

#### Types of Namespaces:
##### 1. Built-in Namespace:

* This contains the names of all built-in functions and variables provided by Python (like len(), print(), int(), etc.).
* It is created when the Python interpreter starts and exists as long as the interpreter is running.
Example:

In [51]:
print(len([1, 2, 3]))  # 'print' and 'len' are built-in functions from the built-in namespace.

3


##### 2. Global Namespace:

* This contains the names defined at the top-level of a script or module (outside of functions and classes).
* Variables declared globally are available throughout the module or file, but not inside functions unless explicitly accessed.

In [52]:
x = 10  # This is in the global namespace

def func():
    print(x)  # Global x is accessible inside the function
func() 

10


##### Local Namespace:

* This contains names defined within a function.
* Variables in this namespace are accessible only inside the function and are destroyed once the function call is complete.

In [54]:
def func():
    y = 5  # Local variable in the local namespace
    print(y)

func()  # Output: 5
print(y)  # Error: y is not defined (local to the function)

5


NameError: name 'y' is not defined

##### Enclosing Namespace:

* This refers to the namespace of an outer function in case of nested functions.
* If a variable is not found in the local namespace, Python will check the enclosing function's namespace (if one exists).

In [55]:
def outer():
    z = 15
    def inner():
        print(z)  # z is from the outer (enclosing) namespace
    inner()

outer()  # Output: 15

15


#### 2. Scope:
* Scope refers to the region or context of a program where a variable or function is accessible.
* A variable’s scope depends on where it is declared, and Python uses the **LEGB** rule to resolve the scope of a variable.

##### The LEGB Rule (Scope Resolution):
This rule defines the hierarchy of namespaces Python searches to find the value of a variable:

* **Local:** Python first looks in the current function’s local scope.
* **Enclosing:** If not found, it looks in the local scope of any enclosing functions (for nested functions).
* **Global:** If the variable is not in the local or enclosing scope, Python checks the global scope.
* **Built-in:** Finally, it checks the built-in namespace for built-in functions and variables.

### Let's create our own module
* create a new .py file (yourName.py) and write some variables and functions and save it
* Also make provisions to check if your module is imported or running standalone
* import it here
* use its functions and variable

In [16]:
# do it here

### DATETIME MODEULE

* The datetime module in Python provides classes for manipulating dates and times. 
* It includes tools to handle date and time arithmetic, comparisons, formatting, and extraction of date/time components.
* Has various classes and methods
* Important classes are **datetime, date ,time and timedelta**

**Key Components of the datetime Module:**
    
* `datetime.date`: Represents a date (year, month, day) without time.
* `datetime.time`: Represents a time (hour, minute, second, microsecond) without date.
* `datetime.datetime`: Combines both date and time.
* `datetime.timedelta`: Represents a duration (difference between two dates or times).
* `datetime.tzinfo`: Handles time zones.

In [68]:
# Create, manipulate and formate dates using datetime class
import datetime

#### datetime class from datetime module

In [65]:


# Creating a date object
today = datetime.date.today()
print("Today's date:", today)

# Creating a custom date object
custom_date = datetime.date(2025, 9, 24)  # year, month, day
print("Custom date:", custom_date)

# Accessing components of a date
print("Year:", today.year)
print("Month:", today.month)
print("Day:", today.day)


Today's date: 2024-09-24
Custom date: 2025-09-24
Year: 2024
Month: 9
Day: 24


In [67]:
help(datetime)

Help on module datetime:

NAME
    datetime - Fast implementation of the datetime type.

MODULE REFERENCE
    https://docs.python.org/3.12/library/datetime.html

    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

CLASSES
    builtins.object
        date
            datetime
        time
        timedelta
        tzinfo
            timezone

    class date(builtins.object)
     |  date(year, month, day) --> date object
     |
     |  Methods defined here:
     |
     |  __add__(self, value, /)
     |      Return self+value.
     |
     |  __eq__(self, value, /)
     |      Return self==value.
     |
     |  __format__(...)
     |      Formats self with strftime.
     |
     |  __ge__(self, value, /)
     |      Return sel

* You'll find methods of this particular class (datetime)

In [66]:
now = datetime.datetime.now()
print("Current date and time:", now)

Current date and time: 2024-09-24 14:50:50.693442


#### Date class
* The instances of date class can be created by  using the keyword **date**

In [71]:
Date1=date(year=2023,month=11,day=6)
print(Date1)
Date2=date.today()
print(Date2)

2023-11-06
2024-09-24


#### Time class
* Each instance carries the details of certain time instant
* Time object is created using the keyword **time** 

In [73]:
Time1=time(hour=22,minute=24,second=10,microsecond=123456)
print(Time1)

22:24:10.123456


In [74]:
help(time)

Help on class time in module datetime:

class time(builtins.object)
 |  time([hour[, minute[, second[, microsecond[, tzinfo]]]]]) --> a time object
 |
 |  All arguments are optional. tzinfo may be None, or an instance of
 |  a tzinfo subclass. The remaining arguments may be ints.
 |
 |  Methods defined here:
 |
 |  __eq__(self, value, /)
 |      Return self==value.
 |
 |  __format__(...)
 |      Formats self with strftime.
 |
 |  __ge__(self, value, /)
 |      Return self>=value.
 |
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |
 |  __gt__(self, value, /)
 |      Return self>value.
 |
 |  __hash__(self, /)
 |      Return hash(self).
 |
 |  __le__(self, value, /)
 |      Return self<=value.
 |
 |  __lt__(self, value, /)
 |      Return self<value.
 |
 |  __ne__(self, value, /)
 |      Return self!=value.
 |
 |  __reduce__(...)
 |      __reduce__() -> (cls, state)
 |
 |  __reduce_ex__(...)
 |      __reduce_ex__(proto) -> (cls, state)
 |
 |  __repr__(self, /)
 |

### datetime.timedelta – Working with Time Differences
* The timedelta class is used for performing arithmetic with dates and times.

In [85]:
import datetime

# Creating timedelta objects
delta = datetime.timedelta(days=5)
print("Delta (5 days):", delta)

today = date.today()

# Adding timedelta to a date
new_date = today + delta
print("Date after 5 days:", new_date)

# Subtracting timedelta from a date
old_date = today - delta
print("Date 5 days ago:", old_date)

custom_date = datetime.date(2023, 9, 24)
print("Custom date:", custom_date)

# Difference between two dates
date_diff = today - custom_date
print("Difference in days:", date_diff.days)


Delta (5 days): 5 days, 0:00:00
Date after 5 days: 2024-09-29
Date 5 days ago: 2024-09-19
Custom date: 2023-09-24
Difference in days: 366


### Formatting and Parsing Dates
* You can format dates and times into strings using the strftime() method and
* convert strings to datetime objects using strptime().

### Strftime function

* It helps to return the string corresponding to the integer arguments present in the datetime object, date object or time object
* Integer to string conversion is done with the help of various format specifiers


In [78]:
Date3=date.today()
print(Date3)
Date33=Date3.strftime('%A,%B,%Y')
print(Date33)

2024-09-24
Tuesday,September,2024


### Format specifiers

| Specifier | Remarks|
|---|----------------------------------------------|
|%d | Returns the day of the month, from 1 to 31.|
|%m| Returns the month of the year, from 1 to 12.|
|%Y| Returns the year in four-digit format (Year with century). like, 2021.|
|%y| Reurns year in two-digit format (year without century). like, 19, 20, 21|
|%A| Returns the full name of the weekday. Like, Monday, Tuesday|
|%a| Returns the short name of the weekday (First three character.). Like, Mon, Tue|
|%B| Returns the full name of the month. Like, June, March|
|%b| Returns the short name of the month (First three character.). Like, Mar, Jun|
|%H| Returns the hour. from 01 to 23.|
|%I| Returns the hour in 12-hours format. from 01 to 12.|
|%M| Returns the minute, from 00 to 59.|
|%S| Returns the second, from 00 to 59.|

| specifier |   Remark   |
|---------|--------------|
|%f| Return the microseconds from 000000 to 999999|
|%p| Return time in AM/PM format|
|%c| Returns a locale’s appropriate date and time representation|
|%x| Returns a locale’s appropriate date representation|
|%X| Returns a locale’s appropriate time representation|
|%z| Return the UTC offset in the form ±HHMM[SS[.ffffff]] (empty string if the object is naive).|
|%Z |Return the Time zone name (empty string if the object is naive).|
|%j |Returns the day of the year from 01 to 366|
|%w |Returns weekday as a decimal number, where 0 is Sunday and 6 is Saturday.|
|%U |Returns the week number of the year (Sunday as the first day of the week) from 00 to 53|
|%W |Returns the week number of the year (Monday as the first day of the week) from 00 to 53|
